File: rust-api/src/schemas/builder.rs
use crate::constants::*;
use crate::error::Result;
use super::types::*;

/// Builder for GenerateParams with method chaining and validation on build.
pub struct GenerateParamsBuilder {
    params: GenerateParams,
}

impl GenerateParamsBuilder {
    /// Create a new builder with the given prompt. All other fields use defaults.
    pub fn new(prompt: impl Into<String>) -> Self {
        Self {
            params: GenerateParams {
                prompt: prompt.into(),
                ..Default::default()
            },
        }
    }

    // -- Non-Option scalar setters -------------------------------------------

    /// Set the generation action (Generate, Img2Img, Infill).
    /// The data-carrying variants include all action-specific fields.
    pub fn action(mut self, action: GenerateAction) -> Self {
        self.params.action = action;
        self
    }

    pub fn model(mut self, model: Model) -> Self {
        self.params.model = model;
        self
    }

    pub fn width(mut self, width: u32) -> Self {
        self.params.width = width;
        self
    }

    pub fn height(mut self, height: u32) -> Self {
        self.params.height = height;
        self
    }

    pub fn steps(mut self, steps: u32) -> Self {
        self.params.steps = steps;
        self
    }

    pub fn scale(mut self, scale: f64) -> Self {
        self.params.scale = scale;
        self
    }

    pub fn cfg_rescale(mut self, cfg_rescale: f64) -> Self {
        self.params.cfg_rescale = cfg_rescale;
        self
    }

    pub fn sampler(mut self, sampler: Sampler) -> Self {
        self.params.sampler = sampler;
        self
    }

    pub fn noise_schedule(mut self, noise_schedule: NoiseSchedule) -> Self {
        self.params.noise_schedule = noise_schedule;
        self
    }

    // -- Option<u64> setter --------------------------------------------------

    pub fn seed(mut self, seed: u64) -> Self {
        self.params.seed = Some(seed);
        self
    }

    // -- Option<Vec<T>> setters ----------------------------------------------

    pub fn characters(mut self, characters: Vec<CharacterConfig>) -> Self {
        self.params.characters = Some(characters);
        self
    }

    /// Set vibes using `Vec<VibeConfig>`, which bundles each vibe item
    /// together with its strength and info_extracted values.
    pub fn vibes(mut self, vibes: Vec<VibeConfig>) -> Self {
        self.params.vibes = Some(vibes);
        self
    }

    // -- Option<struct> setter -----------------------------------------------

    pub fn character_reference(mut self, char_ref: CharacterReferenceConfig) -> Self {
        self.params.character_reference = Some(char_ref);
        self
    }

    // -- Option<String> setters ----------------------------------------------

    pub fn negative_prompt(mut self, prompt: impl Into<String>) -> Self {
        self.params.negative_prompt = Some(prompt.into());
        self
    }

    // -- SaveTarget setters --------------------------------------------------

    /// Set save target to an exact file path.
    pub fn save_path(mut self, path: impl Into<String>) -> Self {
        self.params.save = SaveTarget::ExactPath(path.into());
        self
    }

    /// Set save target to a directory (with optional filename via `SaveTarget::Directory`).
    pub fn save_dir(mut self, dir: impl Into<String>) -> Self {
        self.params.save = SaveTarget::Directory { dir: dir.into(), filename: None };
        self
    }

    // -- Build ---------------------------------------------------------------

    /// Build and validate the params. Returns error if validation fails.
    pub fn build(self) -> Result<GenerateParams> {
        self.params.validate()?;
        Ok(self.params)
    }
}

/// Convenience method on GenerateParams
impl GenerateParams {
    pub fn builder(prompt: impl Into<String>) -> GenerateParamsBuilder {
        GenerateParamsBuilder::new(prompt)
    }
}

File: swift-api/Sources/NovelAIAPI/Schemas/Builder.swift
import Foundation

/// Builder for GenerateParams with method chaining and validation on build
public class GenerateParamsBuilder {
    private var params: GenerateParams

    /// Create a new builder with the given prompt
    public init(prompt: String) {
        self.params = GenerateParams(prompt: prompt)
    }

    // MARK: - Action & Image

    @discardableResult
    public func action(_ action: GenerateAction) -> Self {
        params.action = action
        return self
    }

    @discardableResult
    public func sourceImage(_ image: ImageInput) -> Self {
        params.sourceImage = image
        return self
    }

    @discardableResult
    public func img2imgStrength(_ strength: Double) -> Self {
        params.img2imgStrength = strength
        return self
    }

    @discardableResult
    public func img2imgNoise(_ noise: Double) -> Self {
        params.img2imgNoise = noise
        return self
    }

    // MARK: - Inpaint/Mask

    @discardableResult
    public func mask(_ mask: ImageInput) -> Self {
        params.mask = mask
        return self
    }

    @discardableResult
    public func maskStrength(_ strength: Double) -> Self {
        params.maskStrength = strength
        return self
    }

    @discardableResult
    public func inpaintColorCorrect(_ enabled: Bool) -> Self {
        params.inpaintColorCorrect = enabled
        return self
    }

    // MARK: - Hybrid Mode

    @discardableResult
    public func hybridImg2imgStrength(_ strength: Double) -> Self {
        params.hybridImg2imgStrength = strength
        return self
    }

    @discardableResult
    public func hybridImg2imgNoise(_ noise: Double) -> Self {
        params.hybridImg2imgNoise = noise
        return self
    }

    // MARK: - Characters & References

    @discardableResult
    public func characters(_ characters: [CharacterConfig]) -> Self {
        params.characters = characters
        return self
    }

    @discardableResult
    public func vibes(_ vibes: [VibeItem]) -> Self {
        params.vibes = vibes
        return self
    }

    @discardableResult
    public func vibeStrengths(_ strengths: [Double]) -> Self {
        params.vibeStrengths = strengths
        return self
    }

    @discardableResult
    public func vibeInfoExtracted(_ info: [Double]) -> Self {
        params.vibeInfoExtracted = info
        return self
    }

    @discardableResult
    public func characterReference(_ charRef: CharacterReferenceConfig) -> Self {
        params.characterReference = charRef
        return self
    }

    // MARK: - Prompts

    @discardableResult
    public func negativePrompt(_ prompt: String) -> Self {
        params.negativePrompt = prompt
        return self
    }

    // MARK: - Output Options

    @discardableResult
    public func savePath(_ path: String) -> Self {
        params.savePath = path
        return self
    }

    @discardableResult
    public func saveDir(_ dir: String) -> Self {
        params.saveDir = dir
        return self
    }

    // MARK: - Generation Parameters

    @discardableResult
    public func model(_ model: Model) -> Self {
        params.model = model
        return self
    }

    @discardableResult
    public func width(_ width: Int) -> Self {
        params.width = width
        return self
    }

    @discardableResult
    public func height(_ height: Int) -> Self {
        params.height = height
        return self
    }

    @discardableResult
    public func steps(_ steps: Int) -> Self {
        params.steps = steps
        return self
    }

    @discardableResult
    public func scale(_ scale: Double) -> Self {
        params.scale = scale
        return self
    }

    @discardableResult
    public func cfgRescale(_ cfgRescale: Double) -> Self {
        params.cfgRescale = cfgRescale
        return self
    }

    @discardableResult
    public func seed(_ seed: UInt32) -> Self {
        params.seed = seed
        return self
    }

    @discardableResult
    public func sampler(_ sampler: Sampler) -> Self {
        params.sampler = sampler
        return self
    }

    @discardableResult
    public func noiseSchedule(_ noiseSchedule: NoiseSchedule) -> Self {
        params.noiseSchedule = noiseSchedule
        return self
    }

    // MARK: - Build

    /// Build and validate the params. Throws if validation fails.
    public func build() throws -> GenerateParams {
        try params.validate()
        return params
    }
}

// MARK: - Convenience

extension GenerateParams {
    /// Create a builder with the given prompt
    public static func builder(prompt: String) -> GenerateParamsBuilder {
        return GenerateParamsBuilder(prompt: prompt)
    }
}
